src/Mono.Android/Android.Runtime/DynamicMethodNameCounter.cs

```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Threading;
+
+namespace Android.Runtime {
+	internal class DynamicMethodNameCounter {
+		static int dynamicMethodNameCounter;
+
+		internal static string GetUniqueName ()
+		{
+			if (dynamicMethodNameCounter < int.MaxValue)
+				return Interlocked.Increment (ref dynamicMethodNameCounter).ToString (System.Globalization.CultureInfo.InvariantCulture);
+
+			return Guid.NewGuid ().ToString ();
```

Comment:
I'm not thrilled with any use of `Guid`.

Would it be possible to *separate* constructor from method counting?  We "know" -- or are reasonably sure -- that it's highly unlikely that *methods* will cause an overflow.  *Constructors* can...maybe.  At least, I find it harder to convince myself that Constructors cannot cause an overflow.

I'm inclined to think we should split out method counts from constructor counts.

...and perhaps Constructors should use a `long`?  `Guid` is *really*, ***really***, slow.   `Interlocked.Increment(ref long)` would still be faster, by many orders of magnitude.

Reply:
Or just have long here and keep one counter?

Reply:
Using a `long` would *probably* be fine, but I suspect -- entirely untested! -- that `Interlocked.Increment()` on a `long` will be slower than an `int` on 32-bit hardware.

Then again, how much longer will 32-bit Android devices be a priority?

Reply:
OK, updated to use `long` counter. Hopefully we can *fix* `ConstructorBuilder` soon and get back to `int`.